<a href="https://colab.research.google.com/github/MuthiahAinun/Proyek_Analisis_Sentimen/blob/main/Skema1_Proyek_Analisis_Sentimen_Tsamarah_Muthi'ah_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyek Analisis Sentimen: [Youtube-Comments-dataset]
- **Nama:** [Tsamarah Muthi'ah Abdullah]
- **Email:** [a135xaf486@devacademy.id]
- **ID Dicoding:** [a135xaf48]

# Import Library

In [ ]:
# Menginstall google API
!pip install --upgrade google-api-python-client pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 75.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 78.4 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
  Attempting uninstall: google-api-python-client
    Found existing installation: google-api-python-client 2.160.0
    Uninstalling google-api-python-client-2.160.0:
      Successfully uninstalled google-api-python-client-2.160.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.


In [ ]:
!pip install gensim

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 26.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 39.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 11.5 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
  Attempting uninstall: scipy
    Found existing installation: scipy 1.14.1
    Uninstalling scipy-1.14.1:
      Successfully uninstalled scipy-1.14.1


In [ ]:
# Import Library
import googleapiclient.discovery
import pandas as pd
import time
import random
import re
import string
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Conv1D, GlobalMaxPooling1D, Bidirectional, BatchNormalization
from sklearn.metrics import classification_report, accuracy_score
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, GRU, Conv1D, GlobalMaxPooling1D, Dense, Dropout, Bidirectional, BatchNormalization
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE

# Data Preprocessing dan Pembersihan

In [ ]:
# Membaca dan Menampilkan informasi dataset
dataset = pd.read_csv("NDsO1LT_0lw_youtube_comments.csv")
print("Jumlah data:", len(dataset))

Jumlah data: 85028


In [ ]:
print(dataset.head())

             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  
0  54600  2025-02-08T16:59:31Z  
1      0  2025-03-19T21:51:42Z  
2      0  2025-03-19T21:43:03Z  
3      0  2025-03-19T21:34:32Z  
4      0  2025-03-19T21:20:21Z  


In [ ]:
print("Informasi Dataset:")
print(dataset.info())

Informasi Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85028 entries, 0 to 85027
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   username      84982 non-null  object
 1   text          85028 non-null  object
 2   likes         85028 non-null  int64 
 3   published_at  85028 non-null  object
dtypes: int64(1), object(3)
memory usage: 2.6+ MB
None


In [ ]:
print('\nStatistik Deskriptif:')
print(dataset.describe())


Statistik Deskriptif:
               likes
count   85028.000000
mean        9.727654
std       744.702565
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max    167348.000000


In [ ]:
# Mengecek jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        46
text             0
likes            0
published_at     0
dtype: int64


In [ ]:
# Menghapus nilai kosong
dataset.dropna(inplace=True)

In [ ]:
# Menghapus duplikasi
dataset.drop_duplicates(subset=['text'], inplace=True)

In [ ]:
# Pembersihan teks
def clean_text(text):
    text = text.lower()
    text = ''.join(c for c in text if c.isalnum() or c.isspace())
    return text

dataset['cleaned_text'] = dataset['text'].apply(clean_text)
print("Data setelah pembersihan:")
print(dataset.head())

Data setelah pembersihan:
             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  \
0  54600  2025-02-08T16:59:31Z   
1      0  2025-03-19T21:51:42Z   
2      0  2025-03-19T21:43:03Z   
3      0  2025-03-19T21:34:32Z   
4      0  2025-03-19T21:20:21Z   

                                        cleaned_text  
0  beast games finale drops february 13th go watc...  
1                              saya orang indonesia   
2  مستر بیست عزیز  لطفا دوبله فارسی این ویدیو را ...  
3                    বলদশ আসবন কব পবন জলর কছ দখত চই   
4  this is for girls listen to dalida

In [ ]:
# Mengecek kembali jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        0
text            0
likes           0
published_at    0
cleaned_text    0
dtype: int64


# **2. Ekstraksi Fitur dan Pelabelan Data**

Metode yang digunakan bebas sesuai dengan preferensi masing-masing peserta. Tahapan ini penting untuk mempersiapkan data sehingga dapat diolah lebih lanjut dalam proses pelatihan model.

In [ ]:
# Melabeli komentar secara manual (positif, netral, negatif)
def label_comment(text):
    if any(word in text for word in ["love", "great", "awesome", "good", "nice", "amazing", "fantastic"]):
        return "positif"
    elif any(word in text for word in ["hate", "bad", "awful", "worst", "terrible", "disgusting", "boring"]):
        return "negatif"
    else:
        return "netral"

dataset['label'] = dataset['cleaned_text'].apply(label_comment)
print("Distribusi label:")
print(dataset['label'].value_counts())

Distribusi label:
label
netral     63766
positif     4728
negatif       69
Name: count, dtype: int64


In [ ]:
# Encoding label
label_encoder = LabelEncoder()
dataset['label_encoded'] = label_encoder.fit_transform(dataset['label'])

# **3. Pembangunan Model Deep Learning**

Pilihan algoritma pelatihan ini haruslah sesuai dengan tujuan analisis sentimen yang ingin dicapai.

# Skema 1: LSTM dengan Tokenizer + Padding (80/20)

**Algoritma:** Long Short-Term Memory (LSTM)

**Ekstraksi Fitur:** Tokenizer + Padding

**Alasan Pemilihan:**
- LSTM sangat efektif dalam memproses data berbasis urutan, seperti teks atau kalimat.
- Algoritma ini cocok untuk menangani long-term dependencies dalam urutan kata.
- Tokenizer digunakan untuk mengonversi kata menjadi indeks numerik.
- Padding digunakan untuk memastikan panjang urutan tetap sama agar dapat diproses dalam batch.

In [ ]:
# Tokenisasi teks dengan maksimal 10.000 kata
tokenizer = Tokenizer(num_words=10000, oov_token="<OOV>")
tokenizer.fit_on_texts(dataset['cleaned_text'])

In [ ]:
# Mengubah teks menjadi urutan token
sequences = tokenizer.texts_to_sequences(dataset['cleaned_text'])

In [ ]:
# Padding urutan agar memiliki panjang yang sama
padded = pad_sequences(sequences, maxlen=100, padding='post')

In [ ]:
# Mengatasi Ketidakseimbangan Data dengan SMOTE
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(padded, dataset['label_encoded'])

In [ ]:
# Membagi data menjadi data latih dan data uji (80/20)
X_train, X_test, y_train, y_test = train_test_split(padded, dataset['label_encoded'], test_size=0.2, random_state=42)

In [ ]:
# Membangun Model
print("\nSkema 1: LSTM dengan Tokenizer + Padding (80/20)")

class_weights = {0: 1000, 1: 1, 2: 10}

model1 = Sequential([
    Embedding(10000, 128, input_length=100),
    Bidirectional(LSTM(128, return_sequences=True, dropout=0.2, recurrent_dropout=0.2)),
    GlobalMaxPooling1D(),
    BatchNormalization(),
    Dropout(0.3),
    Dense(128, activation='relu'),
    Dropout(0.3),
    Dense(64, activation='relu'),
    Dropout(0.3),
    Dense(3, activation='softmax')
])


Skema 1: LSTM dengan Tokenizer + Padding (80/20)


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [ ]:
# Mengompilasi model dengan optimizer Adam
model1.compile(loss=tf.keras.losses.SparseCategoricalCrossentropy(), optimizer=tf.keras.optimizers.Adam(learning_rate=0.0001), metrics=['accuracy'])

In [21]:
# Melatih model dengan data latih
model1.fit(X_train, y_train, epochs=30, validation_data=(X_test, y_test), batch_size=64, class_weight=class_weights)

Epoch 1/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 621s 711ms/step - accuracy: 0.7042 - loss: 3.0077 - val_accuracy: 0.6939 - val_loss: 0.9136
Epoch 2/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 637s 727ms/step - accuracy: 0.7162 - loss: 2.0105 - val_accuracy: 0.6921 - val_loss: 0.6790
Epoch 3/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 618s 720ms/step - accuracy: 0.6885 - loss: 1.5884 - val_accuracy: 0.6890 - val_loss: 0.5883
Epoch 4/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 626s 725ms/step - accuracy: 0.7220 - loss: 1.2107 - val_accuracy: 0.7147 - val_loss: 0.5296
Epoch 5/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 624s 727ms/step - accuracy: 0.7228 - loss: 1.1010 - val_accuracy: 0.7704 - val_loss: 0.4286
Epoch 6/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 678s 722ms/step - accuracy: 0.7777 - loss: 0.8334 - val_accuracy: 0.8156 - val_loss: 0.3570
Epoch 7/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 622s 721ms/step - accuracy: 0.8497 - loss: 0.6095 - val_accuracy: 0.9255 - val_loss: 0.1743
Epoch 8/30
858/858 ━━━━━━━━━━━━━━━━━━━━ 619s 722ms/step - accuracy: 0.9092 -

In [24]:
# Evaluasi Model Skema 1
print("\nEvaluasi Model Skema 1: LSTM dengan Tokenizer + Padding (80/20)")

# Prediksi pada data uji
y_pred1 = np.argmax(model1.predict(X_test), axis=1)

# Laporan Klasifikasi
print("\nClassification Report:")
print(classification_report(y_test, y_pred1, target_names=label_encoder.classes_))

# Akurasi
accuracy = accuracy_score(y_test, y_pred1)
print(f"Akurasi Model: {accuracy:.4f}")


Evaluasi Model Skema 1: LSTM dengan Tokenizer + Padding (80/20)
429/429 ━━━━━━━━━━━━━━━━━━━━ 40s 92ms/step

Classification Report:
              precision    recall  f1-score   support

     negatif       0.01      0.60      0.01        10
      netral       1.00      0.93      0.97     12792
     positif       0.97      0.97      0.97       911

    accuracy                           0.94     13713
   macro avg       0.66      0.84      0.65     13713
weighted avg       1.00      0.94      0.97     13713

Akurasi Model: 0.9372


In [28]:
# Inference dan Testing Skema 1
print("\nInference Skema 1: LSTM dengan Tokenizer + Padding (80/20)")
# Fungsi Prediksi Teks
def predict_text_lstm(text, tokenizer, model1, label_encoder):
    # Tokenisasi dan padding teks
    sequence = tokenizer.texts_to_sequences([text])
    padded_sequence = pad_sequences(sequence, maxlen=100, padding='post')

    # Prediksi sentimen
    prediction = model1.predict(padded_sequence)
    label_index = np.argmax(prediction, axis=1)[0]
    label = label_encoder.inverse_transform([label_index])[0]

    print(f"Teks: {text}")
    print(f"Prediksi Sentimen: {label}")

# Prediksi Teks Baru
print("\nPrediksi Teks Baru:")
predict_text_lstm("I love your video!", tokenizer, model1, label_encoder)


Inference Skema 1: LSTM dengan Tokenizer + Padding (80/20)

Prediksi Teks Baru:
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step
Teks: I love your video!
Prediksi Sentimen: positif


**Insight Skema 1:**
1. Model ini memiliki akurasi yang tinggi, mencapai 93.72%. Hal ini menunjukkan bahwa model dapat mengklasifikasikan sebagian besar teks dengan benar.
2. Precision untuk kelas "negatif" sangat rendah (0.01) meskipun recall cukup tinggi (0.60). Ini menunjukkan bahwa model sering salah mengklasifikasikan teks negatif, mungkin karena data negatif sangat sedikit atau model terlalu fokus pada kelas "netral" dan "positif".
3. Precision, recall, dan f1-score pada kelas "netral" dan "positif" sangat tinggi (mendekati 1.00). Hal ini menunjukkan bahwa model mampu mengenali kedua sentimen tersebut dengan sangat baik.
4. Model masih memiliki kelemahan pada kelas negatif, sehingga peningkatan dapat dilakukan dengan teknik augmentasi atau modifikasi arsitektur.
5. Pada teks uji "I love your video!", model berhasil mengklasifikasikannya sebagai "positif", yang merupakan prediksi yang tepat.
